relay-server/src/actors/healthcheck.rs

```diff
@@ -17,16 +18,33 @@ use crate::statsd::RelayGauges;
 
 lazy_static::lazy_static! {
     /// Singleton of the `Healthcheck` service.
-    static ref ADDRESS: RwLock<Option<Addr<HealthcheckMessage>>> = RwLock::new(None);
+    static ref ADDRESS: RwLock<Option<Addr<Healthcheck>>> = RwLock::new(None);
 }
 
-/// Internal wrapper of a message sent through an `Addr` with return channel.
-#[derive(Debug)]
-struct Message<T> {
-    data: T,
-    // TODO(tobias): This is hard-coded to return `bool`.
-    // Might need some associated types to make this work
-    responder: oneshot::Sender<bool>,
+/// Our definition of a service.
+///
+/// Services are much like actors: they receive messages from an inbox and handles them one
+/// by one.  Services are free to concurrently process these messages or not, most probably
+/// should.
+///
+/// Messages always have a response which will be sent once the message is handled by the
+/// service.
+pub trait Service {
+    /// The envelope is what is sent to the inbox of this service.
+    ///
+    /// It is an enum of all the message types that can be handled by this service together
+    /// with the response [sender](oneshot::Sender) for each message.
+    type Envelope: Send + 'static;
```

Comment:
What was the design rationale behind having an enum, instead of making the message type a parameter of `Service`, and then implementing each message handler separately?

For example

```rust
struct MyStruct;

impl Service<Shutdown> for MyStruct;
impl Service<HealthCheck> for MyStruct;
```

Reply:
This came out of the aim to be able to send messages without having to do any boxing to put things in the channel.  Whether that's always desirable that's questionable, but e.g. I think for healthcheck that probably is beneficial.  Really though this should be benchmarked, this version does still allow boxing everything by default quite easily should it turn out to be more performant.

I was initially interested in building both versions to compare, I'm still on the fence myself.  But once this one existed no one disliked this enough to carry on with building the other version.